dynamo-adapter.js

```diff
@@ -0,0 +1,144 @@
+const {promisify} = require('util');
+const {DynamoTable} = require('./slim');
+
+const isCallback = (fn) => typeof fn === 'function';
+
+const validateCall = (methodName, args, minLength) => {
+  if (args.length < minLength)
+    throw new Error(`Incorrect args supplied to method ${methodName}: expected: ${minLength} got: ${args.length}`);
+}
+
+/**
+  Provides an adapter for the DynamoTable class to allow for promisified
+  api usage when omitting the cb parameter.
+*/
+class DynamoAdapter {
+  constructor(name, options) {
+    this._delegate = new DynamoTable(name, options);
+    this.hooks = options.hooks || {};
+
+    // Delegate externally accessed properties
+    Object.keys(this._delegate).forEach(key => {
+      this[key] = this._delegate[key];
+    });
+
+    // NOTE: Rebind all methods so they can be passed without the class instance
+    // Since we only instantiate once per table and maintain the instances,
+    // the performance should be a big concern
+    Object.getOwnPropertyNames(DynamoAdapter.prototype).forEach(key => {
+      if (typeof this[key] === 'function') this[key] = this[key].bind(this);
+    });
+
+    // NOTE: This function needs to be rebound to the delegate for its use case
+    this['mapToDb'] = this._delegate.mapToDb.bind(this._delegate);
+  }
+
+  // Proxies execution to delegate
+  _call(methodName, minLength, ...args) {
+    validateCall(methodName, args, minLength);
+
+    const cb = args[args.length - 1];
+
+    let method = this._delegate[methodName];
+
+    let chain = Promise.resolve(this._runHook(methodName));
+
+    if (!isCallback(cb)) method = promisify(method);
+
+    return chain.then(method.apply(this._delegate, args));
+  }
+
+  _runHook(methodName) {
+    const hook = this.hooks[methodName];
+
+    return hook ? hook() : null;
+  }
+
+  // Allows additional hook for execution BEFORE function
+  addHook(methodName, fn) {
```

Comment:
Added this magical function so that in some cases something can be added instead of completely patching over a function. Its only limited functionality obviously as didn't want to take it too far but just enough to remove a certain painful patch.